include/picongpu/particles/atomicPhysics/kernel/CheckForFieldEnergyOverSubscription.kernel

```diff
@@ -0,0 +1,122 @@
+{
+    /** check all cells for overSubscription
+     *
+     * checks each superCell cell for fieldEnergyUse > fieldEnergy,
+     *  if yes mark cell as oversubscribed and stores rejection probability in rejection probability cache
+     *
+     * @tparam T_numberAtomicPhysicsIonSpecies specialization template parameter used to prevent compilation of all
+     *  atomicPhysics kernels if no atomic physics species is present.
+     */
+    template<uint32_t T_numberAtomicPhysicsIonSpecies>
+    struct CheckForFieldEnergyOverSubscriptionKernel
+    {
+        using VectorIdx = DataSpace<picongpu::simDim>;
+        /** call operator
+         *
+         * called by CheckForOverSubscription atomic physics sub-stage
+         *
+         * @param worker object containing the device and block information, passed by PMACC_KERNEL call
+         * @param areMapping mapping of blockIndex to block superCell index
+         * @param rngFactory factory for uniformly distributed random number generator
+         * @param eFieldBox deviceDataBox giving access to eField values for all local superCells
+         * @param fieldEnergyUseCacheBox deviceDataBox giving access to the field energy use cache for each local
+         * superCell
+         * @param sharedResourcesOverSubscribedBox deviceDataBox giving access to the local shared resources over
+         *  subscription switch for each local superCell
+         * @param rejectionProbabilityCacheCellBox deviceDataBox giving access to localRejectionProbabilityCache for
+         * all cells of all local superCells
+         */
+        template<
+            typename T_Worker,
+            typename T_AreaMapping,
+            typename T_TimeRemainingBox,
+            typename T_EFieldBox,
+            typename T_FieldEnergyUseCacheBox,
+            typename T_SharedRessourcesOverSubscribedBox,
+            typename T_RejectionProbabilityCacheCellDataBox>
+        HDINLINE void operator()(
+            T_Worker const& worker,
+            T_AreaMapping const areaMapping,
+            T_TimeRemainingBox const timeRemainingBox,
+            T_EFieldBox const eFieldBox,
+            T_FieldEnergyUseCacheBox const fieldEnergyUseCacheBox,
+            T_SharedRessourcesOverSubscribedBox sharedResourcesOverSubscribedBox,
+            T_RejectionProbabilityCacheCellDataBox rejectionProbabilityCacheCellBox) const
+        {
+            auto const superCellIdx = KernelIndexation::getSuperCellIndex(worker, areaMapping);
+            auto const superCellFieldIdx
+                = KernelIndexation::getSuperCellFieldIndexFromSuperCellIndex(areaMapping, superCellIdx);
+
+            auto const timeRemaining = timeRemainingBox(superCellFieldIdx);
+            if(timeRemaining <= 0._X)
+                return;
+
+            using FieldEnergyCache = typename T_FieldEnergyUseCacheBox::ValueType;
+            particles::atomicPhysics::localHelperFields::RejectionProbabilityCache_Cell<FieldEnergyCache::numberCells>&
```

Comment:
I wrote it already in an other PR. It looks very dangerous to use a reference here.
If it is a pmacc box please take the return value always by copy. There is only one case where a reference is ok and this is if you use native alpaka to get shared memory.

Reply:
Could be that this should be change in other code parts too.

Reply:
The use of a reference is correct and necessary here, since the dataBox will return a reference to global device memory and we don't want to store a copy of the superCell's rejection probability cache for every worker due to its significant size, ~ 512 * 32-bit.